express-zod-api/src/zts.ts

```diff
@@ -41,48 +58,54 @@ const nodePath = {
   optional: R.path(["questionToken" satisfies keyof ts.TypeElement]),
 };
 
-const onLiteral: Producer = ({ value }: z.ZodLiteral<LiteralType>) =>
-  makeLiteralType(value);
+const onLiteral: Producer = ({ _zod: { def } }: $ZodLiteral) =>
+  f.createUnionTypeNode(
+    def.values.map((entry) =>
+      entry === undefined
+        ? ensureTypeNode(ts.SyntaxKind.UndefinedKeyword)
+        : makeLiteralType(entry),
+    ),
+  );
 
 const onObject: Producer = (
-  { shape }: z.ZodObject<z.ZodRawShape>,
+  { _zod: { def } }: z.ZodObject,
   {
     isResponse,
     next,
     optionalPropStyle: { withQuestionMark: hasQuestionMark },
   },
 ) => {
-  const members = Object.entries(shape).map<ts.TypeElement>(([key, value]) => {
-    const { description: comment, _def } = value as z.ZodType;
-    const isOptional =
-      isResponse && hasCoercion(value)
-        ? value instanceof z.ZodOptional
-        : value.isOptional();
-    return makeInterfaceProp(key, next(value), {
-      comment,
-      isOptional: isOptional && hasQuestionMark,
-      isDeprecated: _def[metaSymbol]?.isDeprecated,
-    });
-  });
+  const members = Object.entries(def.shape).map<ts.TypeElement>(
+    ([key, value]) => {
+      const isOptional =
+        isResponse && hasCoercion(value)
+          ? value instanceof z.ZodOptional
+          : isResponse
+            ? value._zod.qout
+            : value._zod.qin;
```

Comment:
`qin` is not set on ZodOptional.
Filed a fix https://github.com/colinhacks/zod/pull/4153